### PR TITLE
Fix watcher loop quit and channel shouldDrop logic

### PR DESCRIPTION
**Pull request by congqixia**
See also #23393 
/kind bug

### CODE REVIEW

File: internal/datacoord/channel_checker.go [hunk 1]
```diff
@@ -123,7 +123,7 @@ func (c *channelStateTimer) startOne(watchState datapb.ChannelWatchState, channe
 			c.notifyTimeoutWatcher(&ackEvent{ackType, channelName, nodeID})
 			return
 		case <-stop:
-			log.Info("stop timer before timeout",
+			log.Warn("stop timer before timeout",
```
Comment (longjiquan) [inline internal/datacoord/channel_checker.go:126]: Seems that it's normal here, so just keep `Info` level.

**Review by longjiquan — CHANGES_REQUESTED** [review]

File: internal/datacoord/channel_checker.go (hunk 1 above)
Comment (congqixia) [reply internal/datacoord/channel_checker.go:126]: Got it.

File: internal/querynode/shard_segment_detector.go [hunk 2]
```diff
@@ -119,13 +119,16 @@ func (sd *etcdShardSegmentDetector) watch(ch clientv3.WatchChan, collectionID in
 			return
 		case evt, ok := <-ch:
 			if !ok {
-				log.Warn("SegmentDetector event channel closed")
+				log.Warn("SegmentDetector event channel closed, retry...")
+				watchCh := sd.client.Watch(sd.getCtx(), sd.path, clientv3.WithPrefix())
```
Comment (czs007) [inline internal/querynode/shard_segment_detector.go:123]: should check sd.getCtx() is valid?

**Pull request merged**